### ISSUE TITLE
Kubelet plugin registration reliability

### ISSUE THREAD

**Issue by SergeyKanzhelev** [kubernetes/kubernetes#127457]
I am trying to assess the reliability of the kubelet plugins registration.

I am trying it out with the Device Plugin, but the same is likely can be applied to the DRA. The issue describes some findings and concerns, I didn't perform the full review.

I started the e2e to try out things: https://github.com/kubernetes/kubernetes/pull/127304

Plugins documentation: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md

So the flow is:

- Kubelet looks up sockets in `/var/lib/kubelet/plugins_registry/`
- Kubelet tries to get plugin details by calling `GetInfo`
- One information received, kubelet tries to connect to the plugin

- Kubelet proceeds with getting the list of pods to run

The flow is much better in terms of race conditions comparing to the Device Plugin registration when Device Plugin need to detect when kubelet was restarted and reconnect to it.

Two reliability issues I notices immediately in tests:

1. When `GetInfo` fails, there seems to be no retries for while to get plugin details. The timeout for GetInfo is `1sec`. 1 second is a big enough number for an endpoint that simply returns the structure. But flakes are easy to imagine here.
2. When `GetDevicePluginOptions` fails, there is no retry to create a plugin for at least 30 seconds. There may be many reason for a flake when the API didn't return sucessfully for the first time. So retry is essential here.

I need to look deeper into other reliability things:

1. How fast kubelet will detect that the plugin was restarted and will attempt to reconnect? What is the mechanism for it?
2. Will there be retries on `ListAndWatch` failures?
3. Should interface implementing `GetInfo` be checking the health of the plugin? Should the best practice implementation be recreating the socket?

Similar questions can be explored for the DRA.

I think the best way to explore this is to continue working on e2e tests: https://github.com/kubernetes/kubernetes/pull/127304 demonstrating the behavior and a best practice registering the device plugin.

If we can confirm that the kubelet plugins system is almost as reliable as the today's most used device plugin registration mechanism, while eliminating the race condition on kubelet restart (https://github.com/kubernetes/kubernetes/issues/120146#issuecomment-2302666130), we should consider deprecating the `RegisterDevicePluginServer` API.

/sig node
/kind bug

CC: @ffromani @johnbelamaric

**Comment by klueska** [comment]
One thing to note is that unlike the e2e tests which registers the example plugin at `/var/lib/kubelet/plugins_registry/`, actual real-world plugins don't tend to use this mechanism because it came later.

**Comment by SergeyKanzhelev** [comment]
> One thing to note is that unlike the e2e tests which registers the example plugin at `/var/lib/kubelet/plugins_registry/`, many real-world plugins don't tend to use this mechanism because it came later. The NVIDIA GPU plugin being a notable example.

Yes. This makes it harder to reason about the real world reliability of the mechanism that is used by DRA.

**Comment by jackfrancis** [comment]
We've started running some periodic DRA tests on Azure infra via CAPZ and have seen a couple flake types, I don't think they match the above root cause description, but FYI:

- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-provider-azure-conformance-with-ci-artifacts-dra-main/1835039693624840192
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-provider-azure-conformance-with-ci-artifacts-dra-main/1834949095941214208

cc @nojnhuh

**Comment by ffromani** [comment]
/triage accepted
/priority important-longterm

improving the registration process has real-world benefits

**Comment by SergeyKanzhelev** [comment]
@bart0sh pointed on this tests we currently have:

- [Resource Kubelet plugin must register after Kubelet restart](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/dra_test.go#L99)
- [Resource Kubelet plugin must register after plugin restart](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/dra_test.go#L115)
- [Resource Kubelet plugin must retry NodePrepareResources after Kubelet restart](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/dra_test.go#L213)

One concern here that it seems that the design calls for 30 seconds reconnection delay for the plugin. It sounds like a long time for recovery from some flake.

**Comment by lauralorenz** [comment]
I'm going to take a look at this in November unless it was on anyone else's radar. For now,
/assign @lauralorenz

**Comment by jackfrancis** [comment]
Bringing this up to the top, do we have any post-1.32 work to do here?

**Comment by SergeyKanzhelev** [comment]
I started a document with some thoughts on edge cases: https://docs.google.com/document/d/1pQcNqvnd4_DxcGjP334Z34Mj85lM-qTvzq0NrO9kFe0/edit?tab=t.0 Please contribute with more edge cases. Also let me know if you have an idea on a format of a document

**Comment by bart0sh** [comment]
> When GetInfo fails, there seems to be no retries for while to get plugin details. The timeout for GetInfo is 1sec. 1 second is a big enough number for an endpoint that simply returns the structure. But flakes are easy to imagine here.

Do you think we should decrease GetInfo timeout? Can you elaborate on possible flakes? 

Bellow is a quote from the e2e test "Resource Kubelet plugin must recover and register after registration failure" that confirms your observation: GetInfo calls are retried with the timeout 1sec. I tend to believe that this is proved to be working quite good, considering that it's used by CSI plugins for years. However, it can be changed if needed, I guess. We just need a good justification for the change.

```
I0702 15:43:18.643333  469096 nonblockinggrpcserver.go:147] "handling request" logger="DRA kubelet plugin test-driver.cdi.k8s.io.registrar" node="dev" requestID=3 method="/pluginregistration.Registration/GetInfo" request="&InfoRequest{}"
E0702 15:43:18.643488  469096 nonblockinggrpcserver.go:156] "handling request failed" err="simulated GetInfo failure" logger="DRA kubelet plugin test-driver.cdi.k8s.io.registrar" node="dev" requestID=3 method="/pluginregistration.Registration/GetInfo"

I0702 15:43:19.642787  469096 nonblockinggrpcserver.go:147] "handling request" logger="DRA kubelet plugin test-driver.cdi.k8s.io.registrar" node="dev" requestID=4 method="/pluginregistration.Registration/GetInfo" request="&InfoRequest{}"
  E0702 15:43:19.642902  469096 nonblockinggrpcserver.go:156] "handling request failed" err="simulated GetInfo failure" logger="DRA kubelet plugin test-driver.cdi.k8s.io.registrar" node="dev" requestID=4 method="/pluginregistration.Registration/GetInfo"

STEP: unset registration failure mode @ 07/02/25 15:43:20.424

I0702 15:43:21.643428  469096 nonblockinggrpcserver.go:147] "handling request" logger="DRA kubelet plugin test-driver.cdi.k8s.io.registrar" node="dev" requestID=5 method="/pluginregistration.Registration/GetInfo" request="&InfoRequest{}"
  I0702 15:43:21.643541  469096 nonblockinggrpcserver.go:158] "handling request succeeded" logger="DRA kubelet plugin test-driver.cdi.k8s.io.registrar" node="dev" requestID=5 method="/pluginregistration.Registration/GetInfo" response="&PluginInfo{Type:DRAPlugin,Name:test-driver.cdi.k8s.io,Endpoint:/var/lib/kubelet/plugins/test-driver.cdi.k8s.io/dra.sock,SupportedVersions:[v1beta1.DRAPlugin],}"
```

**Comment by bart0sh** [comment]
@SergeyKanzhelev 
> When GetDevicePluginOptions fails, there is no retry to create a plugin for at least 30 seconds. There may be many reason for a flake when the API didn't return sucessfully for the first time. So retry is essential here.

This should be fixed by https://github.com/kubernetes/kubernetes/pull/133335
